src/mappings.ml

```diff
@@ -745,13 +749,17 @@ module Make (M_with_make : M_with_make) : S_with_fresh = struct
         { solver = M.Solver.clone solver; ctx = Stack.copy ctx; last_ctx }
 
       let push { solver; ctx; _ } =
-        let top = Stack.top ctx in
-        Stack.push top ctx;
-        M.Solver.push solver
+        (* TODO: I don't understand why we are duplicating the topmost element of ctx, is this actually the expected behavior? *)
+        match Stack.top_opt ctx with
+        | None -> assert false
+        | Some top ->
+          Stack.push top ctx;
+          M.Solver.push solver
```

Comment:
> (* TODO: I don't understand why we are duplicating the topmost element of ctx, is this actually the expected behavior? *)

Yes, since our context is immutable, when we perform a `push` we duplicated the context so when we later backtrack (with a `pop`) to the context before the push